Scheduler Testing with KWOK (Technical Outcome)

network-charles opened the pull request:
#### What type of PR is this?

/kind documentation
/kind feature

#### What this PR does / why we need it:
This PR describes the technical outcome of using KWOK for scheduler tests. It also contains six scenarios describing these concepts in action:
1. Limit range
2. Node affinity
3. Pod priority and preemption 
4. Pod topology spread constraint
5. Requests and limits
6. Taints and tolerations

#### Which issue(s) this PR fixes:

Referenced issue: #1063

wzshiming reviewed (commented):
Overall, this feels more like a home page or feature index. I thought we could try adding these to the homepage. WDYT?

For the Technical Outcome, I think we can start by narrowing it down to focus on the most typical application scenarios of scheduling. try to find commonalities in the way those items are used, then summarize these practices as technical outcomes

network-charles commented on site/content/en/docs/technical outcomes/simulation.md:
```diff
@@ -0,0 +1,110 @@
+resources. On a 2 GB RAM machine, you can run 100 pods, each with 1 GB of RAM. Simulation in a KWOK cluster differs
+from simulation in a Kind cluster; for example, a Kind cluster deploys real applications inside its cluster. If you
+deploy a nginx pod to a Kind cluster, you could curl to its IP address and get an HTTP response containing its HTML
+page. But in a KWOK cluster, you get nothing because the pod isn't real.
```
I think I understand what you mean now, let's say point 6 in the application and uses cases section which talks about GPUs. We can describe it as a form of “specialized hardware scheduling” where KWOK can be simulated to ensure that certain pods are only scheduled on nodes running GPU hardware. Then the specifications of the GPU nodes were also defined by the developer and then ran on KWOK.

Then in point 2 about “policy enforcement” I can focus on policies that ensure pods are scheduled to nodes with enough resources, specific labels, or taints, preemption of pods, etc. 

Then I follow a similar approach for the rest.

Let me know if that makes sense. Thank you.

wzshiming replied:
Yes, for scheduling, make an article that describes how to simulate various scheduling scenarios, the core intent is for users to read this article to understand what can be done in a scheduling scenario and how to do it. for the others it's similar.

In #1063, I think our definitions of these areas may be a bit off.

In my mind
- Scheduling: we can create nodes and pods, we can configure them with different scheduling policies that meet the needs of all the schedulers.
- Simulation: we can simulate all the normal behavior of kubectl, e.g. `kubectl top`, `kubectl exec`, etc.
- Scalability: we can simulate node scaling.
- Performance: we achieve this by consuming a very small amount of resources, allowing us to simulate large scale tests.

network-charles replied:
Ok, what’s your definition of chaos testing too?

wzshiming replied:
Chaos: we can simulate a variety of user-expected errors to test and improve the robustness of the user's controller.

network-charles replied:
Great, thank you.

network-charles commented:
@wzshiming, I just added a new commit. I still need to update each command specified to point to a URL containing a raw YAML script.

I have a few questions:
- What do you think about these few scenarios I added?
- If you think they are suitable, are there any other scenarios you would like me to add?
- I noticed a scenario that involves installing a custom scheduler in KWOK. See [kube-scheduler-simulator](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/docker-compose.yml). But I don't know how to go about it. Do you think it's worth adding?

Also, I noticed that some use cases that involve installing KWOK into an existing real cluster have a lot to do with scalability, and may best be described in the technical outcome page for scalability.

wzshiming commented:
Here are some typical scheduling scenarios that you can refer to, or even just copy and add to how kwok works with them.

https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/
https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/
https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/
https://kubernetes.io/docs/concepts/policy/limit-range/

network-charles commented:
Great, thank you.

network-charles commented on site/content/en/docs/technical-outcomes/simulation.md:
````diff
@@ -0,0 +1,170 @@
+  - Node 2: 4 CPUs, 4 GB RAM
+<!-- I am going to change each bash command to reference the file via a url. -->
+```bash
+kubectl apply -f node.yaml
````
Hi @wzshiming, how can we specify the URL to the `node.yaml` file and every other referenced file?

I know the URL to the demo path on the main branch is "https://github.com/kubernetes-sigs/kwok/tree/main/demo"

network-charles replied:
Lovely, I will.

network-charles commented:
@wzshiming, I've added all the scenarios. I'm looking forward to your feedback.

wzshiming commented on site/content/en/docs/technical-outcomes/simulation.md:
Rename this file to `site/content/en/docs/technical-outcomes/scheduling/_index.md`
Move `demo/technical-outcomes` to `site/content/en/docs/technical-outcomes/scheduling`

Put them in the same directory and let's focus on scheduling in this one.

wzshiming commented on demo/technical-outcomes/simulation/scenario-1/README.demo:
Run `make demo/technical-outcomes/simulation/scenario-1/README.svg` will generate a svg

network-charles replied:
I keep getting this error:
```bash
~/sigs.k8s.io/kwok$ kubectl apply -f node.yaml
error: the path "node.yaml" does not exist
```

network-charles replied:
Looks like I fixed it. I copied the file to the root directory.

network-charles commented:
@wzshiming, I'm done.

I couldn't [reference other files to be displayed on the page](https://github.com/kubernetes-sigs/kwok/pull/1144#discussion_r1656783537) yet. I got this error:
```bash
Error: Error building site: "/src/site/content/en/docs/technical-outcomes/scheduling/_index.md:1:1": 
closing tag for shortcode 'expand' does not match start tag
```
But I am assuming I can add it and it'll display after the merge. I just wanted to notify you first before I do.

Also, a suggestion, since we have moved that demo folder to here `site/content/en/docs/technical-outcomes/scheduling`, I'll suggest we change the file path from `site/content/en/docs/technical-outcomes/scheduling/technical-outcomes/simulation/` to `site/content/en/docs/technical-outcomes/scheduling/simulation/`. It's shorter this way and removes the redundant folder name, `technical-outcomes`. WDYT?

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-4/README.demo:
```diff
@@ -0,0 +1,40 @@
+sleep 10
+
+## Deploy a pod above the resource limit
+kubectl apply -f pod-beyond-limit.yaml
```
The error output from this command makes the asciinema recorder quit suddenly. We may need to find a way around it or skip adding an `svg` image in this scenario for now.

network-charles commented:
Also, to fix the CI issue, will it be okay to use the below?
```bash
git checkout network-charles-patch-2
git remote add upstream https://github.com/kubernetes-sigs/kwok.git
git fetch upstream
git merge upstream/main
git push
```

wzshiming commented:
> @wzshiming, I'm done.
> 
> I couldn't [reference other files to be displayed on the page](https://github.com/kubernetes-sigs/kwok/pull/1144#discussion_r1656783537) yet. I got this error:
> 
> ```shell
> Error: Error building site: "/src/site/content/en/docs/technical-outcomes/scheduling/_index.md:1:1": 
> closing tag for shortcode 'expand' does not match start tag
> ```
> 
> But I am assuming I can add it and it'll display after the merge. I just wanted to notify you first before I do.
> 
> Also, a suggestion, since we have moved that demo folder to here `site/content/en/docs/technical-outcomes/scheduling`, I'll suggest we change the file path from `site/content/en/docs/technical-outcomes/scheduling/technical-outcomes/simulation/` to `site/content/en/docs/technical-outcomes/scheduling/simulation/`. It's shorter this way and removes the redundant folder name, `technical-outcomes`. WDYT?

I agree with you.

wzshiming commented:
I noticed that you don't seem to do `git rebase' so you can't use code-simple and CI failed.

network-charles commented:
Yes, I don't have to, I can also try “git merge”

The issue here now is that KWOK didn't specify in their contribution guideline what method they support when updating a local branch. But I just found out something now [Kubernetes website doc uses a git rebase](https://kubernetes.io/docs/contribute/new-content/open-a-pr/?t&utm_source=perplexity#changes-from-reviewers) not a git merge.

So it might be okay to use a git rebase here. I tried it and then ran the “make -C site serve” command it ran successfully.

However, we need to open a separate PR for the [contributors page](https://kwok.sigs.k8s.io/docs/contributing/) updating them to use “git rebase” when updating their local branch.

network-charles commented:
What do you suggest we do to fix the CI?

wzshiming commented:
> Error: Error building site: "/opt/build/repo/site/content/en/docs/technical-outcomes/scheduling/_index.md:83:1": failed to extract shortcode: template for shortcode "code-sample" not found
https://app.netlify.com/sites/k8s-kwok/deploys/66825eb087d718000882ccef

network-charles commented:
Building the site stopped working after I changed the file path from `site/content/en/docs/technical-outcomes/scheduling/technical-outcomes/simulation/` to `site/content/en/docs/technical-outcomes/scheduling/simulation/`.

The errors I got:
```bash
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-1/node.yaml" not found
WARN  Expand shortcode is deprecated. Use 'details' instead.
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-1/pod-1.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-1/pod-2.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-2/node.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-2/pod.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-3/node.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-3/no-toleration-pod.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-3/with-toleration-pod.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-4/node.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-4/limit-range.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-4/pod-beyond-limit.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-4/pod-within-limit.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-5/node.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-5/priority-classes.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-5/low-priority-pod.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-5/high-priority-pod.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-6/node-in-us-west-1.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-6/node-in-us-west-2.yaml" not found
ERROR "/docs/technical-outcomes/scheduling/technical-outcomes/simulation/scenario-6/deployment.yaml" not found
```

network-charles commented:
Ah! I just realized that I didn't update the file path to each expanded code on the `_index.md` file.

network-charles commented:
/retest

network-charles commented:
The deploy preview URL didn't show. I was hoping that a test could trigger it to deploy again.

wzshiming commented:
> The deploy preview URL didn't show. I was hoping that a test could trigger it to deploy again.

https://deploy-preview-1140--k8s-kwok.netlify.app/docs/technical-outcomes/scheduling/

wzshiming reviewed (commented):
This page is so big that it gets stuck when I open it, maybe we should put a separate page for each scenario.

network-charles commented:
Ok

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/simulation/scenario-1/README.demo:
I think we should give the scenario a name instead of a sequence number

network-charles replied:
Since we'll have new scenarios for each technical outcome the suggestion makes sense.
I'll use the following:

1. requests-and-limits.md
2. node-affinity.md
3. taints-and-tolerations.md
4. limit-range.md
5. pod-priority-and-preemption.md
6. pod-topology-spread-constraint.md

wzshiming replied:
Also, each scenario can have a separate folder that contains markdown and all its dependencies.

move `requests-and-limits.md` to `requests-and-limits/_index.md`

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/simulation/scenario-1/pod-2.yaml:
```diff
@@ -0,0 +1,20 @@
+      requests:
+        cpu: "1.2"
+      limits:
+          cpu: "1.5"
```
You can run `./hack/update-yaml-format.sh` to format these yaml file

network-charles commented:
Could you let me know your thoughts before I do a git squash and rebase the upstream main branch?

wzshiming commented:
> Could you let me know your thoughts before I do a git squash and rebase the upstream main branch?

If there is no conflict there is no need to rebase. and the git squash is best done before merging.

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range.md:
Let's put all the relevant things together

Rename `docs/technical-outcomes/scheduling/limit-range.md` to `docs/technical-outcomes/scheduling/limit-range/_index.md`, and move `scenario-4` to `docs/technical-outcomes/scheduling/limit-range/`

network-charles replied:
Done

wzshiming replied:
> and move scenario-4 to docs/technical-outcomes/scheduling/limit-range/

You don't seem to have done that.

Ahh. I forgot to mention that the other files are organized in a similar way. put all the relevant things together.

network-charles replied:
Ah, okay. I will do that now.

network-charles replied:
Done

network-charles replied:
Is it okay to move all files in `scenario-4` into `docs/technical-outcomes/scheduling/limit-range/`?

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/_index.md:
```diff
@@ -0,0 +1,49 @@
+---
+title: Simulation Testing
```
```suggestion
title: Scheduling Simulation
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/_index.md:
```diff
@@ -0,0 +1,49 @@
+The cluster can be configured with scheduling policies that meet your scheduler's requirements.
+These scenarios below will be used to describe this:
+
+- [Scenario 1: Scheduling pods with resource requests and limits](/docs/technical-outcomes/scheduling/requests-and-limits)
+- [Scenario 2: Scheduling a pod to a particular node with node-affinity](/docs/technical-outcomes/scheduling/node-affinity)
+- [Scenario 3: Scheduling pods with taints and tolerations](/docs/technical-outcomes/scheduling/taints-and-tolerations)
+- [Scenario 4: Scheduling pods with a limit range](/docs/technical-outcomes/scheduling/limit-range)
+- [Scenario 5: Scheduling pods using pod priority and preemption](/docs/technical-outcomes/scheduling/pod-priority-and-preemption)
+- [Scenario 6: Scheduling pods using pod topology spread constraints](/docs/technical-outcomes/scheduling/pod-topology-spread-constraint)
```
```suggestion
- [Scheduling pods with resource requests and limits](/docs/technical-outcomes/scheduling/requests-and-limits)
- [Scheduling a pod to a particular node with node-affinity](/docs/technical-outcomes/scheduling/node-affinity)
- [Scheduling pods with taints and tolerations](/docs/technical-outcomes/scheduling/taints-and-tolerations)
- [Scheduling pods with a limit range](/docs/technical-outcomes/scheduling/limit-range)
- [Scheduling pods using pod priority and preemption](/docs/technical-outcomes/scheduling/pod-priority-and-preemption)
- [Scheduling pods using pod topology spread constraints](/docs/technical-outcomes/scheduling/pod-topology-spread-constraint)
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/README.demo:
I don't think README is a good name for a demo, maybe `limit-range.demo`

network-charles replied:
Ok, I'll do the same for other files too.

wzshiming commented on site/config.yaml:
```diff
@@ -121,6 +121,12 @@ menu:
+  # Technical Outcomes
+  - identifier: technical-outcomes
+    title: Technical Outcomes
+  - identifier: simulation
+    pageRef: "/docs/technical-outcomes/scheduling"
+    parent: technical-outcomes
```
```suggestion
  - identifier: technical-outcomes-scheduling
    pageRef: "/docs/technical-outcomes/scheduling"
    parent: technical-outcomes
```

wzshiming commented on site/config.yaml:
```diff
@@ -267,6 +273,32 @@ menu:
     pageRef: "/docs/contributing/development"
     parent: contributing
 
+  # Technical Outcome Children
+  - identifier: scenario-1
+    title: "Scenario 1"
+    pageRef: "/docs/technical-outcomes/scheduling/requests-and-limits"
+    parent: simulation
+  - identifier: scenario-2
+    title: "Scenario 2"
+    pageRef: "/docs/technical-outcomes/scheduling/node-affinity"
+    parent: simulation
+  - identifier: scenario-3
+    title: "Scenario 3"
+    pageRef: "/docs/technical-outcomes/scheduling/taints-and-tolerations"
+    parent: simulation
+  - identifier: scenario-4
+    title: "Scenario 4"
+    pageRef: "/docs/technical-outcomes/scheduling/limit-range"
+    parent: simulation
+  - identifier: scenario-5
+    title: "Scenario 5"
+    pageRef: "/docs/technical-outcomes/scheduling/pod-priority-and-preemption"
+    parent: simulation
+  - identifier: scenario-6
+    title: "Scenario 6"
+    pageRef: "/docs/technical-outcomes/scheduling/pod-topology-spread-constraint"
+    parent: simulation
```
```suggestion
  # Technical Outcome Scheduling Children
  - identifier: technical-outcomes-scheduling-requests-and-limits
    title: "Requests and Limits"
    pageRef: "/docs/technical-outcomes/scheduling/requests-and-limits"
    parent: technical-outcomes-scheduling
  - identifier: technical-outcomes-scheduling-requests-and-limits
    title: "Node Affinity"
    pageRef: "/docs/technical-outcomes/scheduling/node-affinity"
    parent: technical-outcomes-scheduling
  - identifier: technical-outcomes-scheduling-taints-and-tolerations
    title: "Taints and Tolerations"
    pageRef: "/docs/technical-outcomes/scheduling/taints-and-tolerations"
    parent: technical-outcomes-scheduling
  - identifier: technical-outcomes-scheduling-limit-range
    title: "Limit Range"
    pageRef: "/docs/technical-outcomes/scheduling/limit-range"
    parent: technical-outcomes-scheduling
  - identifier: technical-outcomes-scheduling-pod-priority-and-preemption
    title: "Pod Priority and Preemption"
    pageRef: "/docs/technical-outcomes/scheduling/pod-priority-and-preemption"
    parent: technical-outcomes-scheduling
  - identifier: technical-outcomes-scheduling-pod-topology-spread-constraint
    title: "Pod Topology Spread Constraint"
    pageRef: "/docs/technical-outcomes/scheduling/pod-topology-spread-constraint"
    parent: technical-outcomes-scheduling
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/_index.md:
```diff
@@ -0,0 +1,49 @@
+title: Simulation Testing
+---
+
+# Simulation Testing with KWOK
+
+{{< hint "info" >}}
+
+This document walks you through the technical outcome of using KWOK for simulation testing.
+
+{{< /hint >}}
+
+This documentation describes the technical outcome of using KWOK for simulation testing in your environment.
+It goes over the key simulation features of using KWOK, and then provides a step-by-step guide to creating
+various scheduling scenarios using KWOK.
+
+## Key Simulation Features of KWOK
+
+1. **Lifecycle configuration:** The pod, node, and node lease can be simulated.
+
+    - **Pod:** Status information like Pending, ContainerCreating, Running, and Terminating can be simulated.
+    - **Node:** Status information like NodeReady, and Ready states can be simulated.
+
+2. **Kubelet server**
+
+    - **Metrics:** Using the [Kubernetes Metric Server](https://kwok.sigs.k8s.io/docs/examples/metrics-server/), or [Prometheus](https://kwok.sigs.k8s.io/docs/examples/prometheus/), metrics can be simulated in KWOK.
+    - **Exec:** Commands can be simulated to be [executed](https://kwok.sigs.k8s.io/docs/user/exec-configuration/) inside a container running on a single fake pod or multiple pods.
+    - **Log:** Container [logs](https://kwok.sigs.k8s.io/docs/user/logs-configuration/) can be simulated in a single pod or multiple pods.
+    - **Attach:** The input/output streams from the main process running in a container can be [attached](https://kwok.sigs.k8s.io/docs/user/attach-configuration/#attach-configuration) to KWOK.
+    This can be simulated in a single pod or multiple pods.
+    - **PortForward:** [Port forwarding](https://kwok.sigs.k8s.io/docs/user/port-forward-configuration/#clusterportforward), which establishes a direct connection to a container port, can be simulated
+    for a single pod or multiple pods.
```
I think this part should be move to a new page `site/content/en/docs/technical-outcomes/_index.md`

network-charles replied:
Ok, that also means that other technical outcomes may have to be referenced here too, correct?

network-charles replied:
Since the aim is to have several pages introducing how to use KWOK combined with various scenarios, we can remove Lines 5 - 32.
This page focuses on scheduling not simulation.
There will be no need to move the page.

wzshiming replied:
Well, it's a wise choice, I also think it's a bit of a duplicate of the home page

network-charles replied:
True, I will update it now.

network-charles replied:
Also, we could update the title to "Scheduler Testing", since the scheduling process is real, using "Scheduling Simulation" as the title is confusing.

network-charles replied:
What do you think?

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/_index.md:
I saw this created the demo file, but what didn't generate svg?

network-charles replied:
Yes, I outlined the reason in this [review](https://github.com/kubernetes-sigs/kwok/pull/1140#discussion_r1658523887).

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/_index.md:
```diff
@@ -0,0 +1,114 @@
+---
+title: "Scenario 4"
```
```suggestion
title: "Limit Range"
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/_index.md:
```diff
@@ -0,0 +1,114 @@
+title: "Scenario 4"
+---
+
+# Scenario 4: Scheduling pods with a limit range
```
```suggestion
# Scheduling pods with a limit range
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+## Deploy a pod above the resource limit
+
+- Pod specification: 0.7 CPU
```
```suggestion
- Pod specification: 700m CPU
```

I think we should use a unified CPU unit `m`

network-charles replied:
Ok

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+## Deploy a pod within the resource limit
+
+- Pod specification: 0.4 CPU
```
```suggestion
- Pod specification: 400m CPU
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/limit-range.yaml:
```diff
@@ -0,0 +1,11 @@
+  name: cpu-resource-constraint
+spec:
+  limits:
+  - max:
+      cpu: "0.5"
+    min:
+      cpu: "0.1"
```
```suggestion
  - max:
      cpu: 500m
    min:
      cpu: 100m
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/node.yaml:
```diff
@@ -0,0 +1,31 @@
+    memory: 4Gi
+    pods: "55"
+  capacity:
+    cpu: "4"
```
```suggestion
    cpu: 4
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/pod-beyond-limit.yaml:
```diff
@@ -0,0 +1,14 @@
+    - containerPort: 80
+    resources:
+      requests:
+        cpu: "0.7"
```
```suggestion
        cpu: 500m
```

network-charles replied:
I think you meant
```suggestion
        cpu: "700m"
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/pod-within-limit.yaml:
```diff
@@ -0,0 +1,16 @@
+    image: fake-image
+    ports:
+    - containerPort: 80
+    resources:
+      requests:
+        cpu: "0.4"
+      limits:
+        cpu: "0.4"
```
```suggestion
    resources:
      requests:
        cpu: 400m
      limits:
        cpu: 400m
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/_index.md:
```diff
@@ -0,0 +1,88 @@
+---
+title: "Scenario 2"
```
```suggestion
title: "Node Affinity"
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/_index.md:
```diff
@@ -0,0 +1,88 @@
+title: "Scenario 2"
+---
+
+# Scenario 2: Scheduling a pod to a particular node with node-affinity
```
```suggestion
# Scheduling a pod to a particular node with node-affinity
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/README.demo:
Rename to node-affinity.demo

network-charles replied:
Ok

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/README.demo:
```diff
@@ -0,0 +1,40 @@
+## View clusters
+kwokctl get clusters
+
+## Deploy node
+kubectl apply -f node.yaml
```
```suggestion
# Create a node
kwokctl scale node --replicas 1
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/_index.md:
````diff
@@ -0,0 +1,114 @@
+{{< /expand >}}
+
+```bash
+kubectl apply -f node.yaml
````
```suggestion
kwokctl scale node --replicas 1
```

network-charles replied:
This means that I'll have to delete `node.yaml` and also remove its `expand` file reference, correct?

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/limit-range/_index.md:
````diff
@@ -0,0 +1,114 @@
+kwokctl get clusters
+```
+
+## Deploy node
+
+- Node: 1 worker node in the cluster
+
+{{< expand "node.yaml" >}}
+
+{{< code-sample file="node.yaml" >}}
+
+{{< /expand >}}
````
```suggestion
## Create nodes
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/_index.md:
````diff
@@ -0,0 +1,88 @@
+```bash
+kubectl apply -f node.yaml
+```
+
+{{< expand "node.yaml" >}}
+
+{{< code-sample file="node.yaml" >}}
+
+{{< /expand >}}
````
```suggestion
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/_index.md:
````diff
@@ -0,0 +1,88 @@
+- Nodes: 2 worker nodes in the cluster
+
+```bash
+kubectl apply -f node.yaml
````
```suggestion
kwokctl scale node --replicas 2
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/_index.md:
````diff
@@ -0,0 +1,88 @@
+kwokctl get clusters
+```
+
+## Deploy nodes
+
+Below are the node resource specifications:
+
+- Nodes: 2 worker nodes in the cluster
````
```suggestion
## Create nodes
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/_index.md:
````diff
@@ -0,0 +1,88 @@
+kubectl label node node-1 region=us-west-2
+```
+
+## Pod to be scheduled
+
+The pod has a node affinity configured with a key value of “region=eu-west-2”. This matches the label on node-1.
+
+## Scheduling process
+
+### Step 1: Deploy pod
+
````
```suggestion
## Scheduling process

### Step 1: Deploy pod

The pod has a node affinity configured with a key value of `region=eu-west-2`. This matches the label on node-1.
```

wzshiming reviewed (commented):
I left a few suggested changes, similar changes are needed for the other cases

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/node.yaml:
Remove this file

network-charles replied:
I just realized that we needed a custom node running at a maximum CPU capacity of 4. So I am creating this file again.

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/pod-priority-and-preemption.demo:
```diff
@@ -0,0 +1,40 @@
+kwokctl get clusters
+
+## Deploy node
+kubectl apply -f node.yaml
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/pod-priority-and-preemption.demo:
```diff
@@ -0,0 +1,40 @@
+kubectl apply -f node.yaml
+
+## Wait for a while so the node is running
+sleep 5
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/pod-priority-and-preemption.demo:
```diff
@@ -0,0 +1,40 @@
+
+## Deploy a high-priority pod
+kubectl apply -f high-priority-pod.yaml
+
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/pod-priority-and-preemption.demo:
```diff
@@ -0,0 +1,40 @@
+kubectl apply -f high-priority-pod.yaml
+
+## Wait for a while so the pods are scheduled
+sleep 5
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/pod-priority-and-preemption.demo:
```diff
@@ -0,0 +1,40 @@
+kubectl get pod
+
+## Wait for a while so you can observe the result
+sleep 10
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/pod-priority-and-preemption.demo:
```diff
@@ -0,0 +1,40 @@
+kubectl describe pod high-priority-pod | awk '/Events:/,/pod to node/'
+
+## Wait for a while so you can observe the result
+sleep 10
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/_index.md:
````diff
@@ -0,0 +1,107 @@
+kwokctl get clusters
+```
+
+## Deploy node
+
+Below are the node resource specifications:
+
+- Node: 4 worker nodes in the cluster
+  - Node 1: 
+    - Key = topology.kubernetes.io/region
+    - Value = us-west-1
+  - Node 2: 
+    - Key = topology.kubernetes.io/region
+    - Value = us-west-1
+  - Node 3:
+    - Key = topology.kubernetes.io/region
+    - Value = us-west-2
+  - Node 4:
+    - Key = topology.kubernetes.io/region
+    - Value = us-west-2
+
+{{< expand "node-in-us-west-1.yaml" >}}
+
+{{< code-sample file="node-in-us-west-1.yaml" >}}
+
+{{< /expand >}}
+
+{{< expand "node-in-us-west-2.yaml" >}}
+
+{{< code-sample file="node-in-us-west-2.yaml" >}}
+
+{{< /expand >}}
+
+```bash
+kubectl apply -f node-in-us-west-1.yaml
+kubectl apply -f node-in-us-west-2.yaml
+```
````
ditto

wzshiming replied:
Let use `kubectl label` to label each node

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/node-in-us-west-1.yaml:
Remove this file

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/node-in-us-west-2.yaml:
Remove this file

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/pod-topology-spread-constraint.demo:
```diff
@@ -0,0 +1,29 @@
+## View clusters
+kwokctl get clusters
+
+## Deploy nodes
+kubectl apply -f node-in-us-west-1.yaml
+kubectl apply -f node-in-us-west-2.yaml
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/pod-topology-spread-constraint.demo:
```diff
@@ -0,0 +1,29 @@
+kubectl apply -f node-in-us-west-2.yaml
+
+## Wait for a while so the nodes are running
+sleep 5
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/pod-topology-spread-constraint.demo:
```diff
@@ -0,0 +1,29 @@
+sleep 5
+
+## Create deployment
+kubectl apply -f deployment.yaml
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/pod-topology-spread-constraint.demo:
```diff
@@ -0,0 +1,29 @@
+kubectl apply -f deployment.yaml
+
+## Wait for a while so the pods are scheduled
+sleep 5
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/pod-topology-spread-constraint.demo:
```diff
@@ -0,0 +1,29 @@
+kubectl get pod -o wide
+
+## Wait for a while so you can observe the result
+sleep 10
```
ditto

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/_index.md:
```diff
@@ -0,0 +1,106 @@
+title: "Scenario 1"
+---
+
+# Scenario 1: Scheduling pods with resource requests and limits
```
```suggestion
# Scheduling pods with resource requests and limits
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/_index.md:
```diff
@@ -0,0 +1,106 @@
+---
+title: "Scenario 1"
```
```suggestion
title: "Scheduling pods with resource requests and limits"
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/_index.md:
````diff
@@ -0,0 +1,106 @@
+kwokctl get clusters
+```
+
+## Deploy nodes
+
+Below are the node resource specifications:
+
+- Nodes: 2 worker nodes in the cluster
+  - Node 1: 4 CPUs
+  - Node 2: 2 CPUs
+
+{{< expand "node.yaml" >}}
+
+{{< code-sample file="node.yaml" >}}
+
+{{< /expand >}}
+
+```bash
+kubectl apply -f node.yaml
+```
````
ditto

network-charles replied:
I ended up replacing `##` with `#`.

wzshiming reviewed (commented):
- Replace `##` with `#` in `.demo` files.
- Remove all sleep in `.demo` files.
- Add prefix `${WORK_DIR}/` to `.yaml` file path in `.demo` files.

There are a lot of similar changes that I think you can make.

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/low-priority-pod.yaml:
```diff
@@ -9,6 +9,6 @@ spec:
     resources:
       requests:
-        cpu: "3"
+        cpu: 300m
```
I believe you meant `3000m`. I have updated this and the other.

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/node.yaml:
Remove this file

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/node.yaml:
Remove this file

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/_index.md:
```diff
@@ -0,0 +1,104 @@
+title: "Scenario 6"
+---
+
+# Scenario 6: Scheduling pods using pod topology spread constraints
```
```suggestion
# Scheduling pods using pod topology spread constraints
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/pod-topology-spread-constraint/_index.md:
```diff
@@ -0,0 +1,104 @@
+---
+title: "Scenario 6"
```
```suggestion
title: "Scheduling pods using pod topology spread constraints"
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/taints-and-tolerations/_index.md:
```diff
@@ -0,0 +1,97 @@
+title: "Scenario 3"
+---
+
+# Scenario 3: Scheduling Pods with taints and tolerations
```
```suggestion
# Scheduling Pods with taints and tolerations
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/taints-and-tolerations/_index.md:
```diff
@@ -0,0 +1,97 @@
+---
+title: "Scenario 3"
```
```suggestion
title: "Scheduling Pods with taints and tolerations"
```

wzshiming commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/requests-and-limits.demo:
```diff
@@ -0,0 +1,26 @@
+# View cluster
+kwokctl get clusters
+
+# Deploy node
+kubectl apply -f ${WORK_DIR}/node.yaml
```
```suggestion
# Create nodes
kwokctl scale node node-2c --replicas 1 --param '.allocatable.cpu="2000m"'
kwokctl scale node node-4c --replicas 1 --param '.allocatable.cpu="4000m"'
```

wzshiming reviewed (commented):
There are still a few tweaks that need to be made, and I've left a comment and I feel like it's almost done, merged all the commits.

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/_index.md:
````diff
@@ -0,0 +1,121 @@
+## Create a node
+
+```bash
+kubectl apply -f node.yaml
````
Ok

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/pod-priority-and-preemption/node.yaml:
Nice

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/requests-and-limits.demo:
```diff
@@ -0,0 +1,27 @@
+kwokctl get clusters
+
+# Create nodes
+kwokctl scale node node-2c --replicas 1 --param '.allocatable.cpu="2000m"'
+kwokctl scale node node-4c --replicas 1 --param '.allocatable.cpu="4000m"'
```
I believe you meant

```suggestion
kwokctl scale node node-1 --replicas 1 --param '.allocatable.cpu="4000m"'
kwokctl scale node node-2 --replicas 1 --param '.allocatable.cpu="2000m"'
```

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/_index.md:
````diff
@@ -0,0 +1,106 @@
+{{< /expand >}}
+
+```bash
+kubectl apply -f node.yaml
````
The node names should be none-1 and node-2

```suggestion
kwokctl scale node node-1 --replicas 1 --param '.allocatable.cpu="4000m"'
kwokctl scale node node-2 --replicas 1 --param '.allocatable.cpu="2000m"'
```

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/requests-and-limits/_index.md:
```diff
@@ -0,0 +1,106 @@
+  - Node 1: 4 CPUs
+  - Node 2: 2 CPUs
+
+{{< expand "node.yaml" >}}
+
+{{< code-sample file="node.yaml" >}}
+
+{{< /expand >}}
```
```suggestion
```

network-charles commented on site/content/en/docs/technical-outcomes/scheduling/node-affinity/node.yaml:
I've updated this to
```bash
kwokctl scale node node-1 --replicas 1 --param '.allocatable.cpu="4000m"'
kwokctl scale node node-2 --replicas 1 --param '.allocatable.cpu="2000m"'
```

wzshiming commented:
/ok-to-test
/release-note-none

network-charles commented:
I'll resolve the error.

network-charles commented:
/retest

wzshiming commented:
/override pull-kwok-e2e-test-main

wzshiming commented:
/approve
/lgtm
